File: camera/ogg_reader.go
package main

import (
	"fmt"
	"io"
)

// ReadOggOpusPackets reads an OGG bitstream from r and calls onPacket
// for each Opus audio packet extracted. The first two pages (OpusHead
// and OpusTags headers) are skipped automatically.
//
// This is a minimal OGG page parser — just enough to extract Opus
// packets from ffmpeg's `-f ogg -c:a libopus` output. It does not
// validate CRCs or handle chained streams.
func ReadOggOpusPackets(r io.Reader, onPacket func(data []byte)) error {
	headerPages := 0

	for {
		packets, err := readOggPage(r)
		if err == io.EOF {
			return nil
		}
		if err != nil {
			return err
		}

		// Skip the first two pages (OpusHead + OpusTags).
		if headerPages < 2 {
			headerPages++
			continue
		}

		for _, pkt := range packets {
			if len(pkt) > 0 {
				onPacket(pkt)
			}
		}
	}
}

// readOggPage reads one OGG page and returns the extracted packets.
func readOggPage(r io.Reader) ([][]byte, error) {
	// OGG page header is 27 bytes fixed.
	var hdr [27]byte
	if _, err := io.ReadFull(r, hdr[:]); err != nil {
		return nil, err
	}

	if hdr[0] != 'O' || hdr[1] != 'g' || hdr[2] != 'g' || hdr[3] != 'S' {
		return nil, fmt.Errorf("ogg: bad capture pattern: %x", hdr[:4])
	}

	numSegments := int(hdr[26])

	// Read segment table (one byte per segment, value = segment length).
	segTable := make([]byte, numSegments)
	if numSegments > 0 {
		if _, err := io.ReadFull(r, segTable); err != nil {
			return nil, fmt.Errorf("ogg: reading segment table: %w", err)
		}
	}

	// Read all segment data.
	var totalSize int
	for _, s := range segTable {
		totalSize += int(s)
	}
	data := make([]byte, totalSize)
	if totalSize > 0 {
		if _, err := io.ReadFull(r, data); err != nil {
			return nil, fmt.Errorf("ogg: reading page data: %w", err)
		}
	}

	// Reassemble packets from segments. A segment < 255 bytes
	// terminates a packet. A segment of exactly 255 bytes means
	// the packet continues in the next segment.
	var packets [][]byte
	var current []byte
	offset := 0
	for _, segLen := range segTable {
		current = append(current, data[offset:offset+int(segLen)]...)
		offset += int(segLen)
		if segLen < 255 {
			packets = append(packets, current)
			current = nil
		}
	}
	// Partial packet (last segment was 255) spans to next page — discard.
	// Not expected for Opus at 20ms frame size (<255 bytes per frame).

	return packets, nil
}

File: camera/ghostcam/ogg_reader.py
"""Minimal OGG page parser → Opus packets.

Mirrors camera/ogg_reader.go. Just enough to extract Opus packets from
ffmpeg's `-f ogg -c:a libopus` output. No CRC validation; no chained
streams.

Async because the page bytes arrive from an asyncio.StreamReader that
wraps the ffmpeg side-channel pipe (see capture.py for how the pipe is
set up — the `pipe:{wfd}` URL pattern from Spike 2).
"""

from __future__ import annotations

import asyncio
import logging
from collections.abc import Awaitable, Callable

logger = logging.getLogger(__name__)


async def read_ogg_opus_packets(
    stream: asyncio.StreamReader,
    on_packet: Callable[[bytes], Awaitable[None] | None],
) -> None:
    """Read pages until EOF, dispatching each Opus packet.

    Skips the first two pages (OpusHead + OpusTags). on_packet may be
    sync or async — both forms are awaited correctly.
    """
    header_pages = 0
    while True:
        try:
            packets = await _read_one_page(stream)
        except asyncio.IncompleteReadError:
            return  # EOF
        except OggError as e:
            logger.debug("ogg parse error: %s", e)
            return

        if header_pages < 2:
            header_pages += 1
            continue

        for pkt in packets:
            if not pkt:
                continue
            result = on_packet(pkt)
            if asyncio.iscoroutine(result):
                await result


class OggError(ValueError):
    pass


async def _read_one_page(stream: asyncio.StreamReader) -> list[bytes]:
    hdr = await stream.readexactly(27)
    if hdr[:4] != b"OggS":
        raise OggError(f"bad capture pattern: {hdr[:4]!r}")
    num_segments = hdr[26]
    seg_table = await stream.readexactly(num_segments) if num_segments else b""
    total = sum(seg_table)
    data = await stream.readexactly(total) if total else b""

    packets: list[bytes] = []
    current = bytearray()
    offset = 0
    for seg_len in seg_table:
        current.extend(data[offset:offset + seg_len])
        offset += seg_len
        if seg_len < 255:
            packets.append(bytes(current))
            current = bytearray()
    return packets
